01-overview-of-ml-systems.md

```diff
@@ -0,0 +1,110 @@
+| Data                   | Often Static                                                 | Constantly shifting                                    |
+| Fairness               | Often not a focus                                            | Must be considered (7)                                 |
+| Interpretability       | Often not a focus, performance is the north start (8)        | Must be considered                                     |
+
+(1) No focus on model deployment and maintenance.
+
+(2) MLEs, Sales teams, Product Managers, ML platform teams all have different requirements for the same system.
+
+(3) Most modern systems batch up queries at inference time. In research, it doesn't matter if the time to answer a single query is high as long as the throughput of predictions for the whole test dataset is acceptable. In production, the time to answer a single query is important, so we need to balance that with the throughput.
+
+(4)  `...once you deploy your model into the real world, latency matters a lot. In 2017, an Akamai study found that a 100 ms delay can hurt conversion rates by 7%. 20 In 2019, Booking.com found that an increase of about 30% in latency cost about 0.5% in conversion rates—“a relevant cost for our business.” 21 In 2016, Google found that more than half of mobile users will leave a page if it takes more than three seconds to load. 22 Users today are even less patient.` 
+
+(5) Latency must be monitored using percentiles, not averages. Don't ignore higher percentiles like p90 through p99 as the most valuable customers, tend to be the ones with more data and the ones that experience slower inferences.
+
+(6) Consider using high percentiles like p90 to specify the latency requirements of the system.
+
+(7) ML systems encode the past and perpetuate the biases the training data has. To make things worse, ML systems can discriminate at scale potentially affecting millions of people. 
+
+(8) For example, *ensemble models* are common in research but very uncommon in production because of their low interpretability and high complexity.
+
+
+
+# ML systems vs traditional software
+On the surface, bringing SWE best practices to ML engineering is a good idea. However, it is not that simple. SWE systems keep data and code separate by design and the efforts of versioning, testing, deploying and monitoring are focused on the  **code** artefacts.  
+
+ML systems are part code, part data and part artefacts created from the two. **All of these need to be versioned,  tested, deployed and monitored.**
+
+Challenges:
+
+- How to version data?
+- How to test that a data sample is good or bad for the system?
+- How to deploy a trained model with millions of parameters that requires gigabytes of RAM? How to do so if it needs to be deposed on the edge?
+- How to monitor and debug deployed models considering they may be hard to interpret?
```

Comment:
In the age of LLMs, these questions have changed


- How to version data?
  - Still a problem
- How to test that a data sample is good or bad for the system?
  - For Transformers with tokenized text / image / video I think you can just through all the data at it. There is no bad data.
- How to deploy a trained model with millions of parameters that requires gigabytes of RAM? How to do so if it needs to be deposed on the edge?
  - LLM training does not run on edge. But edge inferrance is still a problem.
- How to monitor and debug deployed models considering they may be hard to interpret?
  - LLMs are easy to interpret. Or we can build another LLM to interpret the LLM, and then another LLM to compare the results. But it's probably less of a problem for LLM.